### ISSUE TITLE
Current schema allows `ObjectComponent` to be an output of multiple `CodeRun`s

### ISSUE THREAD

**Issue by nathan-cummings-ukaea** [FAIRDataPipeline/data-registry#117]
This is a mistake and presents a problem for describing provenance. An `ObjectComponent` should only be an output of a single `CodeRun`.

To fix this, we'll likely need to change the model definitions to something like:

```
class ObjectComponent(BaseModel):
    ...
    outputs_of = models.ForeignKey(CodeRun, on_delete=models.PROTECT, related_name='outputs', blank=True)
    ...

class CodeRun(BaseModel):
    ...
    EXTRA_DISPLAY_FIELDS = ('outputs')
    ...
```

This way, keeping the names the same should hopefully not break any of the language-specific APIs, but if @alahiff could weigh in on whether that is true, and let me know where else in the code would need to be changed that would be good.

**Comment by nathan-cummings-ukaea** [comment]
Currently, we have:
```
class ObjectComponent(BaseModel):
    ...
    EXTRA_DISPLAY_FIELDS = ('outputs_of')
    ...

class CodeRun(BaseModel):
    ...
    outputs = models.ManyToManyField(ObjectComponent,  related_name='outputs_of', blank=True)
    ...
```

**Comment by nathan-cummings-ukaea** [comment]
Ahh... so the solution to this is to use a lazy import:
```
class ObjectComponent(BaseModel):
    ...
    outputs_of = models.ForeignKey('CodeRun', on_delete=models.PROTECT, related_name='outputs', blank=True)
    ...

class CodeRun(BaseModel):
    ...
    EXTRA_DISPLAY_FIELDS = ('outputs')
    ...
    inputs = models.ManyToManyField(ObjectComponent, related_name='inputs_of', blank=True)
    ...
```

**Comment by alahiff** [comment]
I had a look into this and ended up with the same as the last comment, except I would have called it `output_of` rather than `outputs_of`.

The only possible issue is that when you define an `ObjectComponent` you have to specify the `CodeRun` that it is an output of. Is this ok? i.e. will a `CodeRun` always be created first before the output `ObjectComponent`s are created?

**Comment by nathan-cummings-ukaea** [comment]
> I had a look into this and ended up with the same as the last comment, except I would have called it `output_of` rather than `outputs_of`.

I suggested keeping the name the same in case that reduced any impact on the language API codes. If it doesn't make any difference, `output_of` should be the name of course.

**Comment by richardreeve** [comment]
I'm not sure that `CodeRun`s are always created during the `initialise()` step - @soniamitchell or @mjb3 may be able to comment. If not, this sounds like something that may be disruptive though unfortunately, so may be best left till after the paper is finished...

On a related note, I should say that not all `ObjectComponent`s are created by code runs though (some are `register`ed separately), so can we create an `ObjectComponent` without a `CodeRun` at all under this scheme? If not, it's not going to work.

**Comment by nathan-cummings-ukaea** [comment]
Yeah the outputs field will be optional. It's just that if you are going to fill it, the `CodeRun` will need to exist first.

**Comment by soniamitchell** [comment]
I always create a CodeRun entry in `initialise`. I create object components in `finalise`, then PATCH their URLs to the CodeRun entry. Sometimes I DELETE CodeRun entries in `finalise`.

**Comment by richardreeve** [comment]
Okay, well maybe this is possible then...

**Comment by mjb3** [comment]
Hi Richard, me too re CodeRun being created at initialise. As you mentioned there are things like registering issues that I assumed might happen outside a code run so I guess that means objects (and therefore OC's) can be created outside of CodeRuns..?

**Comment by soniamitchell** [comment]
@mjb3 I forgot about issues! You’re right, they’re created outside of code runs.

**Comment by mjb3** [comment]
Yeah, I just couldn't quite remember if issue is an object or its own stand alone thing (I can't seem to access the schema image any more for some reason)

**Comment by soniamitchell** [comment]
From memory, issues are attached to object components. The components can be associated with data products, scripts, configs, and maybe other stuff too. The scripts and configs are registered during `initialise`. Hopefully that’s accurate, I’m not near my laptop.

**Comment by richardreeve** [comment]
That seems right. In that case it doesn't matter that the issues are being created separately, so long as the object components themselves are created after code run creation...

**Comment by richardreeve** [comment]
@alahiff I'm adding this to the 1.0 milestone because it sounds like this won't be disruptive (or not too disruptive!) after all...
